modules/light-clients/14-avalanche/client_state.go

```diff
@@ -283,8 +342,335 @@ func IsMatchingClientState(subject, substitute ClientState) bool {
+		return errorsmod.Wrap(err, "updated client state failed basic validation")
+	}
+
+	// The new consensus state is merely used as a trusted kernel against which headers on the new
+	// chain can be verified. The root is just a stand-in sentinel value as it cannot be known in advance, thus no proof verification will pass.
+	// The timestamp and the NextValidatorsHash of the consensus state is the blocktime and NextValidatorsHash
+	// of the last block committed by the old chain. This will allow the first block of the new chain to be verified against
+	// the last validators of the old chain so long as it is submitted within the TrustingPeriod of this client.
+	// NOTE: We do not set processed time for this consensus state since this consensus state should not be used for packet verification
+	// as the root is empty. The next consensus state submitted using update will be usable for packet-verification.
+	newConsState := NewConsensusState(
+		avaUpgradeConsState.Timestamp,
+		avaUpgradeConsState.Vdrs,
+		avaUpgradeConsState.StorageRoot,
+		avaUpgradeConsState.SignedStorageRoot,
+		avaUpgradeConsState.ValidatorSet,
+		avaUpgradeConsState.SignedValidatorSet,
+		avaUpgradeConsState.SignersInput,
+	)
+
+	setClientState(clientStore, cdc, newClientState)
+	SetConsensusState(clientStore, cdc, newConsState, newClientState.LatestHeight)
+	setConsensusMetadata(ctx, clientStore, avaUpgradeClient.LatestHeight)
+
+	return nil
+}
+
+// VerifyMembership is a generic proof verification method which verifies a proof of the existence of a value at a given CommitmentPath at the specified height.
+// The caller is expected to construct the full CommitmentPath from a CommitmentPrefix and a standardized path (as defined in ICS 24).
+// If a zero proof height is passed in, it will fail to retrieve the associated consensus state.
+func (cs ClientState) VerifyMembership(
+	ctx sdk.Context,
+	clientStore sdk.KVStore,
+	cdc codec.BinaryCodec,
+	height exported.Height,
+	delayTimePeriod uint64,
+	delayBlockPeriod uint64,
+	proof []byte,
+	path exported.Path,
+	value []byte,
+) error {
+	if cs.GetLatestHeight().LT(height) {
+		return errorsmod.Wrapf(
+			ibcerrors.ErrInvalidHeight,
+			"client state height < proof height (%d < %d), please ensure the client has been updated", cs.GetLatestHeight(), height,
+		)
+	}
+
+	if err := verifyDelayPeriodPassed(ctx, clientStore, height, delayTimePeriod, delayBlockPeriod); err != nil {
+		return err
+	}
+
+	consensusState, found := GetConsensusState(clientStore, cdc, height)
+	if !found {
+		return errorsmod.Wrap(clienttypes.ErrConsensusStateNotFound, "please ensure the proof was constructed against a height that exists on the client")
+	}
+
+	vdrs, totalWeigth, err := ValidateValidatorSet(ctx, consensusState.Vdrs)
+	if err != nil {
+		return err
+	}
+
+	chainID, _ := ids.ToID([]byte(cs.ChainId))
+	unsignedMsg, _ := warp.NewUnsignedMessage(
+		chainID,
+		ids.Empty,
+		consensusState.ValidatorSet,
+	)
+
```

Comment:
Can we add some comments here, please? Like "Check that 2/3rds of validators signed validator set" (simple statement of what we are verifying). Same for the storage root `Verify()` call which, from our discussions, I understand to be the signed merkle root used for proof verification.